### PR TITLE
Adding easychallenge code

### PR DESCRIPTION
**Pull request by mauriiciobarbosa**
https://user-images.githubusercontent.com/4331095/128789977-c0cfc6fc-7f79-4def-8338-092043a785c6.mov

### CODE REVIEW

File: lib/components/button.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    required this.onPressed,
+  });
+
+  final String text;
+  final VoidCallback? onPressed;
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.symmetric(vertical: 8.0),
```
Comment (williamhjcho) [inline lib/components/button.dart]: ElevatedButton tem o style onde você consegue botar o padding (e o constraints do SizedBox abaixo)

Se você estiver usando esse padding pra espaçamento entre componentes, então deveria estar na tela, não no componente em si, a menos que ele tenha uma definição bem clara de comportamento

File: lib/components/input_field.dart [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  final String label;
+  final String hint;
+  final TextInputFormatter? inputFormatter;
+  final TextEditingController inputFieldController;
+  final Function(String)? onChanged;
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.symmetric(vertical: 8.0),
```
Comment (williamhjcho) [inline lib/components/input_field.dart]: Mesmo ponto acima sobre o padding em componentes

File: lib/components/input_field.dart [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      child: TextField(
+        textAlign: TextAlign.center,
+        controller: inputFieldController,
+        keyboardType: keyboardType,
+        inputFormatters: inputFormatter != null ? [inputFormatter!] : null,
```
Comment (williamhjcho) [inline lib/components/input_field.dart]:
```dart
[
  if (inputFormatter != null) inputFormatter,
]
```

File: lib/data/dtos/investment_simulation_response.dart [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  InvestmentSimulationResult toDomain() {
+    initializeDateFormatting('pt_BR');
+
+    final moneyFormatter =
+        NumberFormat.currency(locale: 'pt_BR', symbol: 'R\$');
+    final percentFormatter =
+        NumberFormat.decimalPercentPattern(locale: 'pt_BR', decimalDigits: 2);
+    final dateFormatter = DateFormat('dd/MM/yyyy', 'pt_BR');
```
Comment (williamhjcho) [inline lib/data/dtos/investment_simulation_response.dart]: A gente tinha falado antes, mas esse deveria ser um modelo puro, sem localização nem nada

File: lib/data/uri_parser.dart [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    String path,
+    InvestmentSimulationRequest request,
+  ) {
+    try {
+      String params = '?investedAmount=${request.investedAmount}'
+          '&index=${request.index}'
+          '&rate=${request.rate}'
+          '&isTaxFree=${request.isTaxFree}'
+          '&maturityDate=${request.maturityDate}';
+      return Uri.parse(baseUrl + path + params);
```
Comment (williamhjcho) [inline lib/data/uri_parser.dart]: Cria Uri na mão ou usa o `replace()` pra adicionar os query params
editar string na mão é bem error-prone

File: lib/main.dart [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          ),
+        ),
+      ),
+      home: BlocProvider(
+        create: (BuildContext context) => InvestmentCubit(
+          repository: new InvestmentSimulationRepository(),
+        ),
+        child: FormScreen(
+          onPressed: (childContext) {
+            Navigator.push(
```
Comment (williamhjcho) [inline lib/main.dart]: ao invés de fazer push de uma rota diretamente, você pode registrar elas no `MaterialApp` também

```dart
routes: {
  '/some-route': (context) => MyWidget(), 
}
```

File: lib/components/button.dart (hunk 1 above)
Comment (mauriiciobarbosa) [reply lib/components/button.dart]: boa. Como faria pra adiciona o SizedBox?

File: lib/components/input_field.dart (hunk 2 above)
Comment (mauriiciobarbosa) [reply lib/components/input_field.dart]: image
Se eu adiciono o `contentPadding` os campos ficam bem próximos. =/

File: lib/components/button.dart (hunk 1 above)
Comment (mauriiciobarbosa) [reply lib/components/button.dart]: style tem um minimumSize, pra substituir o SizedBox. Investigar!
Mover padding pra fora.

File: lib/components/input_field.dart (hunk 2 above)
Comment (mauriiciobarbosa) [reply lib/components/input_field.dart]: mesmo ponto acima.

File: lib/data/dtos/investment_simulation_response.dart (hunk 4 above)
Comment (mauriiciobarbosa) [reply lib/data/dtos/investment_simulation_response.dart]: * colocar formatters na tela.
* verificar depois a questão do locale pra pegar do settings. (pensar em utilizar com um bloc externo)

File: lib/data/uri_parser.dart (hunk 5 above)
Comment (mauriiciobarbosa) [reply lib/data/uri_parser.dart]: verificar o que eu fiz.

File: lib/data/uri_parser.dart [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import 'dtos/investment_simulation_request.dart';
+
+class UriParser {
+  Uri https(
+    String baseUrl,
+    InvestmentSimulationRequest request,
+  ) {
+    try {
+      final uri = Uri.parse(baseUrl);
+      return uri.replace(queryParameters: request.toJson());
```
Comment (mauriiciobarbosa) [inline lib/data/uri_parser.dart:10]: @williamhjcho precisei tornar todos os campos do `InvestmentSimulationRequest` Strings pra fazer o `replace` funcionar. Sabe dizer o motivo?

File: lib/screens/form/form_screen.dart [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            ResultContainer.routeName,
+            arguments: ResultArguments(state.amount, state.rate, state.date),
+          );
+        },
+      ),
+    );
+  }
+}
+
+class FormScreen extends StatelessWidget {
```
Comment (mauriiciobarbosa) [inline lib/screens/form/form_screen.dart:31]: @williamhjcho o ideal seria tratar o `Form` como um `StatefulWidget`? Percebi que os valores dos campos não são mantidos num hot reload, por exemplo.

File: test/screens/result/result_screen_test.dart [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+        widget,
+        'Simular novamente',
+        isEnabled: true,
+      ),
+    );
+    await tester.tap(button);
+    await tester.pumpAndSettle();
+  });
+
+  testWidgets('when back button is clicked', (tester) async {
```
Comment (mauriiciobarbosa) [inline test/screens/result/result_screen_test.dart:301]: @williamhjcho por algum motivo o teste falha quando clico no voltar. Saberia dizer o motivo?

> `══╡ EXCEPTION CAUGHT BY GESTURE ╞═══════════════════════════════════════════════════════════════════
The following NoSuchMethodError was thrown while handling a gesture:
Closure call with mismatched arguments: function 'main.<anonymous closure>.<anonymous closure>'
Receiver: Closure: (Object?) => void
Tried calling: main.<anonymous closure>.<anonymous closure>()
Found: main.<anonymous closure>.<anonymous closure>(Object?) => void
When the exception was thrown, this was the stack:
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:54:5)
#1      _objectNoSuchMethod (dart:core-patch/object_patch.dart:97:9)
#2      Function._apply (dart:core-patch/function_patch.dart:11:30)
#3      Function.apply (dart:core-patch/function_patch.dart:34:12)
#4      _ExpectedFunction._run (package:test_api/src/frontend/expect_async.dart:193:23)
#5      _ExpectedFunction.max6 (package:test_api/src/frontend/expect_async.dart:176:7)
#6      _ExpectedFunction.max1 (package:test_api/src/frontend/expect_async.dart:144:41)
#7      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:989:21)
#8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
#9      TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:607:11)
#10     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:296:5)
#11     BaseTapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:230:7)
#12     PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:475:9)
#13     PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:93:12)
#14     PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:138:9)
#15     _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:397:8)
#16     PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:136:18)
#17     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:122:7)
#18     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:439:19)
#19     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:419:22)
#20     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:287:11)
#21     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:374:7)
#22     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:338:5)
#23     TestWidgetsFlutterBinding.handlePointerEvent (package:flutter_test/src/binding.dart:507:13)
#24     WidgetTester.sendEventToBinding.<anonymous closure> (package:flutter_test/src/widget_tester.dart:792:15)
#25     WidgetTester.sendEventToBinding.<anonymous closure> (package:flutter_test/src/widget_tester.dart:791:39)
#28     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#29     WidgetTester.sendEventToBinding (package:flutter_test/src/widget_tester.dart:791:27)
#30     TestGesture.up.<anonymous closure> (package:flutter_test/src/test_pointer.dart:392:24)
#31     TestGesture.up.<anonymous closure> (package:flutter_test/src/test_pointer.dart:390:39)
#34     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:71:41)
#35     TestGesture.up (package:flutter_test/src/test_pointer.dart:390:27)
#36     WidgetController.tapAt.<anonymous closure> (package:flutter_test/src/controller.dart:278:21)
<asynchronous suspension>
<asynchronous suspension>(elided 5 frames from dart:async and package:stack_trace)
Handler: "onTap"
Recognizer:
  TapGestureRecognizer#84d81`

File: lib/data/uri_parser.dart (hunk 7 above)
Comment (williamhjcho) [reply lib/data/uri_parser.dart:10]: query params só aceita string, 
isso é da definição da URL

File: lib/screens/form/form_screen.dart (hunk 8 above)
Comment (williamhjcho) [reply lib/screens/form/form_screen.dart:31]: não é pq você não tá passando os initial values do state management pros componentes também? (ou pros controllers)

File: lib/screens/form/form_screen.dart [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  FormScreen({required this.onPressed});
+
+  final Function(BuildContext context) onPressed;
```
Comment (williamhjcho) [inline lib/screens/form/form_screen.dart]: callback que precisa de context?

e é bom especificar toda a assinatura da função (o retorno também)

File: test/screens/result/result_screen_test.dart (hunk 9 above)
Comment (williamhjcho) [reply test/screens/result/result_screen_test.dart:301]: me parece que tá falando que os types tão mismatched (das funções/callbacks)

File: lib/screens/form/form_screen.dart (hunk 10 above)
Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart]: eu preciso do contexto pra navegar pra próxima tela:

> onPressed: (context) {
          final state = BlocProvider.of<InvestmentCubit>(context).state
              as InvestmentInitialState;
          Navigator.pushNamed(
            context,
            ResultContainer.routeName,
            arguments: ResultArguments(state.amount, state.rate, state.date),
          );
        },
        
  Se utilizar o contexto do `FormContainer` eu tomo esse erro:
  `BlocProvider.of() called with a context that does not contain a InvestmentCubit.`

Ou teria uma forma mais simples de ter acesso ao contexto?

File: lib/screens/form/form_screen.dart (hunk 8 above)
Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart:31]: boa, faz sentido. Criei o método `_fillFieldsIfNeeded(state as InvestmentInitialState);` pra atualizar os controllers em toda mudança de configuração.

File: lib/screens/form/form_screen.dart [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                    },
+                  ),
+                ),
+                Padding(
+                  padding: const EdgeInsets.symmetric(vertical: 8.0),
+                  child: InputField(
+                    keyboardType: TextInputType.number,
+                    label: 'Qual a data de vencimento do investimentos? *',
+                    hint: 'dia/mês/ano',
+                    // inputFormatter: MaskTextInputFormatter(mask: "##/##/####"),
```
Comment (mauriiciobarbosa) [inline lib/screens/form/form_screen.dart]: @williamhjcho alguma dica pra adicionar `inputFormatter` nos campos de data e percentual? Tentei usar o `MaskTextInputFormatter`, [dessa lib](https://pub.dev/packages/mask_text_input_formatter), mas toda vez que tento atualizar o valor do campo ele limpa o último valor digitado. =/

https://user-images.githubusercontent.com/4331095/131255903-2a83bbcd-64f1-43d8-947e-a941a16743c1.mov

File: test/screens/result/result_screen_test.dart (hunk 9 above)
Comment (mauriiciobarbosa) [reply test/screens/result/result_screen_test.dart:301]: puts, verdade. Como o `onBack` é um `VoidCallback` o correto seria utilizar o `expectAsync0(() {}, count: 1)` ao invés do `expectAsync1((_) {}, count: 1)`. Boa!

File: lib/screens/form/form_screen.dart [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return BlocBuilder<InvestmentCubit, InvestmentCubitState>(
+      builder: (context, state) {
+        _fillFieldsIfNeeded(state as InvestmentInitialState);
+        return Scaffold(
+          appBar: AppBar(
+            title: Text('Formulário'),
+          ),
+          body: SingleChildScrollView(
```
Comment (mauriiciobarbosa) [inline lib/screens/form/form_screen.dart:60]: @williamhjcho tem alguma forma simples de fazer os campos ficarem ao centro da tela? Talvez ocupando um percentual dela seria interessante. Atualmente ele tá assim.

Screen Shot 2021-08-29 at 14 18 13.

Comment (williamhjcho) [reply lib/screens/form/form_screen.dart:60]: o próprio InputField não tem anda associado a TextAlignment ou coisa similar?

File: lib/screens/form/form_screen.dart (hunk 11 above)
Comment (williamhjcho) [reply lib/screens/form/form_screen.dart]: Essa lib aí precisa que o mask seja salvo no state, e não seja recriado toda vez pq fica guardando estados internos na instancia dele

File: lib/screens/form/form_screen.dart (hunk 10 above)
Comment (williamhjcho) [reply lib/screens/form/form_screen.dart]: hmm o mais simples aqui seria mudar o onPressed pra `void Function(XXX)` onde XXX é o state que você quer ou os dados dele
tipo `void Function(InvestmentInitialState)` ou  `void Function(double amount, double rate, ...)` o que achar que seja melhor.

---

o `of()` tem que prestar atenção na hierarquia e contexts que tão sendo acessados mesmo

```dart
Widget build(BuildContext context) {  // context 1
  return BlocProvider<InvestmentCubit>( // provido aqui
      create: (BuildContext context) => InvestmentCubit(), // context 2
      child: FormScreen( // é child direta do provider, mas olha o context abaixo
        onPressed: () { //se não recebesse o context aqui
          // context 1 - esse context tá acima do provider, por isso que não vai achar
          final state = BlocProvider.of<InvestmentCubit>(context).state 
              as InvestmentInitialState;
          Navigator.pushNamed(
            context, // context 1 - tem um navigator acima desse widget, então funciona ok
            ResultContainer.routeName,
            arguments: ResultArguments(state.amount, state.rate, state.date),
          );
        },
      ),
    );
}
```

Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart]: hm, boa. Ficou claro o motivo de não achar o Bloc então. Vou tomar cuidado com esses `context`. Por hora segui com a sugestão de passar os dados pra próxima tela via parâmetros, daí não preciso da instância do cubit nesse listener.

File: lib/screens/form/form_screen.dart (hunk 11 above)
Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart]: Boa mano, realmente não tinha percebido. Toda vez que emitia um novo valor no cubit ele era reiniciado. Daí mantive as instâncias no widget.

`final _dateFormatter = MaskTextInputFormatter(mask: "##/##/####");`
`final _rateFormatter = MaskTextInputFormatter(mask: "###");`
  
Você acha que faz sentido?

https://user-images.githubusercontent.com/4331095/131830087-72619e63-0cf6-49e1-9e6d-3bd97bbc2d09.mov

File: lib/screens/form/form_screen.dart (hunk 12 above)
Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart:60]: Ele até tem um `textAlign` e um `textAlignVertical`, mas acredito que, pelo fato de o conteúdo da tela estar dentro de um `SingleChildScrollView` esses parâmetros não têm efeito, muito menos o `mainAxisAlignment` do `Column`, que está logo abaixo do `SingleChildScrollView`. Faz sentido?

Comment (mauriiciobarbosa) [reply lib/screens/form/form_screen.dart:60]: Vou fechar esse PR, que já tá muito grande, e ajusto isso em outro PR. Valeu d+ pelo help até aqui Cho,

**Pull request merged**